automation/deploy_igz/deploy_remote.py

```diff
@@ -0,0 +1,293 @@
+            )
+
+    @staticmethod
+    def _get_image_tag(branch, tag):
+        return f"v{tag}-{branch}"
+
+    @staticmethod
+    def _execute_local_proc_interactive(cmd):
+        proc = subprocess.Popen(
+            cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, text=True
+        )
+        for line in proc.stdout:
+            yield line
+        proc.stdout.close()
+        ret_code = proc.wait()
+        if ret_code:
+            raise subprocess.CalledProcessError(ret_code, cmd)
+
+    def _exec_local(self, cmd: List[str], live=False) -> str:
+        logger.debug("Exec local: %s", " ".join(cmd))
+        buf = io.StringIO()
+        for line in self._execute_local_proc_interactive(cmd):
+            buf.write(line)
+            if live:
+                print(line, end="")
+        output = buf.getvalue()
+        return output
+
+    def _exec_remote(self, cmd: List[str], live=False):
+        cmd_str = " ".join(cmd)
+
+        logger.debug("Exec remote: %s", cmd_str)
+
+        stdin_stream, stdout_stream, stderr_stream = self._ssh_client.exec_command(
+            cmd_str
+        )
+
+        stdout = ""
+        if live:
+            while True:
+                line = stdout_stream.readline()
+                stdout += line
+                if not line:
+                    break
+                print(line, end="")
+        else:
+            stdout = stdout_stream.read().decode("utf8")
+
+        stderr = stderr_stream.read().decode("utf8")
+
+        exit_status = stdout_stream.channel.recv_exit_status()
+
+        if exit_status:
+            raise RuntimeError(
+                f"Command '{cmd_str}' finished with failure ({exit_status})\n{stderr}"
+            )
+
+        return stdout
+
+    def _get_current_version(self) -> str:
+        sys.path.append(os.path.dirname(os.path.abspath(__file__)) + "/..")
+
+        from version import version_file
+
+        return str(version_file.read_unstable_version_prefix())
+
+    def _find_current_git_branch(self) -> str:
+        cmd = ["git", "rev-parse", "--abbrev-ref", "HEAD"]
+        return self._exec_local(cmd).strip()
+
+    def _make_mlrun_api(self, image_tag):
+        logger.info("Building mlrun-api docker image")
+        os.environ["MLRUN_VERSION"] = image_tag
+        os.environ["MLRUN_DOCKER_REPO"] = self._config["DOCKER_REGISTRY"]
+        cmd = ["make", "api"]
+        return self._exec_local(cmd, live=True).strip()
+
+    def _connect_to_node(self, node):
+        logger.debug(f"Connecting to {node}")
+
+        self._ssh_client = paramiko.SSHClient()
+        self._ssh_client.set_missing_host_key_policy(paramiko.WarningPolicy)
+        self._ssh_client.connect(
+            node,
+            username=self._config["USER"],
+            password=self._config["PASSWORD"],
+        )
+
+    def _disconnect_from_node(self):
+        self._ssh_client.close()
+
+    def do_replacing(self):
```

Comment:
This method is way to long IMHO.
I'd also call it `patch_mlrun_resources`
and add private methods for: `_push_image_to_docker_registry`, `_configure_deployment_image_pull_policy`, `_configure_deployment_image` and `_wait_for_deployment_readiness`